src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+// \ int extra
+//
+
+struct UnitTypeInfo {
+
+  UnitTypeInfo(UnitType type, UnitTypeStats baseStats)
+      : type(type),
+        baseStats(baseStats) {
+    stats = baseStats;
+  }
+
+  UnitType type;
+  UnitTypeStats baseStats;   // immutable
+  UnitTypeStats stats;       // mutable
+
+  std::vector<Modification> modifications;
+
+  void AddModification(Modification mod) {
+
+    bool samePriorityAsLast = modifications.back().priority = mod.priority;
+    if (samePriorityAsLast) {
+      // this opt is not so important, could be removed
+      modifications.push_back(mod);
+      mod.ApplyToUnit(&stats, &baseStats);
+
+    } else {
+      modifications.push_back(mod); // TODO: instert at the correct position based on the priority
+
+      // re-eval from the start
+      stats = baseStats;
+      for (auto i : modifications) {
+        i.ApplyToUnit(&stats, &baseStats);
+      }
+    }
+  }
+
+  void MakeCurrentStastTheBaseStats() { // for converting (in order to hide the info about the opponent upgrades)
+    modifications.clear();
+    baseStats = stats;
+  }
+
+  // Gamelogic looks only at the stats.
+  // Gui looks in baseStats.X for base values and
+  // stat.X - baseStats.X for the bonus from technologies and
+  // for more complex info at the modification list.
+};
+
+/// BuidlingTypeInfo, CivilizationInfo, TechnologyInfo
+/// - a template thing
+/// - copy paste
+
+/// Note: The type, stats member are stored without a pointer and contain no pointers,
+///       so there is no indirection in accessing every field. To minimize indirection
+///       the two class should be stored in different vectors again without a pointer
+///       eg. vector<UnitTypeInfo> unitInfos;
+
+
+struct ObjectTypeStats {
+
+  float fireRate;
+  Damage damage;
+
+  float minRange;
+  float maxRange;
+
+  float accuracy; // TODO: how many types of accuracy are there ?
+  bool projectileSpeed;
+
+  int maxHp;
+  Armor armor;
+  float lineOfSight;
+
+  float creationTime;
+  ResourceAmount cost;
+
+  enum class GarrisonType {
+    None,
+    Units, // TransportShip ?
+    VillagersAndFootArchersAndInfantry, // TownCenter and towers
+    NonSiege, // Castle
+    Villagers,  // Khmer houses
+    /// Units from production buildings can be trained directly into garrison,
+    /// but not garrison from outside.
+    Production,
+    /// Same as Production plus Relics can garrison
+    ProductionAndRelics
+  };
+
+  GarrisonType garrisonType;
+  int garrisonCapacity;
+
+  /// Based on the type:
+  /// - Villager, FishingShip: gather rate
+  /// - Production building: production speed
+  /// - Other: ?
+  float workRate;
+
+  int conversionResistanceLevel;
+
+  // Used only by resource spot and animals
+  ResourceAmount resources;
+};
+
+constexpr int sizeof_ObjectTypeStats = sizeof(ObjectTypeStats); 
+// 320 = 2 * sizeof(Damage) + 80
+// 200 types * 2 per info * 320 = 125K
+
+struct UnitTypeStats : public ObjectTypeStats {
+
+  float radius;
+  float speed;
```

Comment:
> // NOTE: This population check will not account for zero-population units such as sheep,
    //       but that should be fine.

Add `population` field which will have 1 for units and 0 for animals and other things

Reply:
Also consider the Karambit Warrior of the Malai, which takes up 0.5 population. This could either be implemented by multiplying all population values by 2, and dividing them by 2 for display, or by using float (but that might perhaps end up storing something like 39.499 population then)

Reply:
... The floats will be bad **, if there are only the 1, .5 and 0 we could have an enum {One, Half, None} to avoid the *2 in the game data and have the PlayerStats have an internal *2 int but with the getter returning the value/2 (to encapsulate the *2 fully inside the PlayerStats).

** **Irrelevant 1**: If a remember correctly form university (I dont know how to search for it) , I believe .5f and 1f addition only up to a upper limit could not produce an error because they are both powers of 2. (but still I don't thing floats is a good idea)

**Irrelevant 2**: Its interesting to image in the original game dev team a "Game Designer" saying to the "Coder": "You know how every unit for the past 20 years has an integer population value, how about a 0.5 ?"